[HUDI-6086]. Improve HiveSchemaUtil#generateCreateDDL With ST.

slfan1989 opened the pull request:
### Change Logs

The code of HiveSchemaUtil#generateCreateDDL uses a lot of append, which makes the code very difficult to read. Usually, in this case, we should use antlr's ST to generate SQL. This jira will use ST to improve this part of the code.

### Impact

none.

### Risk level (write none, low medium or high below)

none.

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

slfan1989 commented:
@danny0405 Can you help review this pr? Thank you very much!

danny0405 commented on hudi-sync/hudi-hive-sync/src/main/java/org/apache/hudi/hive/util/HiveSchemaUtil.java:
```diff
@@ -65,6 +67,34 @@ public class HiveSchemaUtil {
+      + "<" + ROW_FORMAT + ">\n"
+      + "<" + LOCATION_BLOCK + ">"
+      + "TBLPROPERTIES (\n"
+      + "<" + PROPERTIES + ">)";
```
two concerns here:
1. Where does the antlr jar comes from? Do we need to package antlr jar explicitly
2. the antlr grammar template is not that straight-forward, should refactoring like using the stringBuilder with params be enough now?

slfan1989 replied:
@danny0405 Thank you very much for helping review the code!

1. Antlr was introduced into the Hudi project in [HUDI-4111] Bump ANTLR runtime version in Spark 3.x (#5606), we can find the reference of antlr in the parent pom.xml
```
<antlr.version>4.8</antlr.version>
```
`hudi-hive-sync` references `hive-exec`, so we can use `antlr` directly.

2. Thank you for your question! I agree with your point of view, in some scenarios, the antlr grammar template is not that straight-forward. But for generating sql, I think we can use it, because the template can better describe the components of sql.

HiveSchemaUtil#CREATE_TABLE_TEMPLATE
```
private static final String CREATE_TABLE_TEMPLATE =
      "CREATE <" + EXTERNAL + ">TABLE <if(" + DATABASE_NAME + ")>`<" + DATABASE_NAME + ">`.<endif>"
      + "`<" + TABLE_NAME + ">`(\n"
      + "<" + LIST_COLUMNS + ">)\n"
      + "<" + PARTITIONS + ">\n"
      + "<" + BUCKETS + ">\n"
      + "<" + ROW_FORMAT + ">\n"
      + "<" + LOCATION_BLOCK + ">"
      + "TBLPROPERTIES (\n"
      + "<" + PROPERTIES + ">)";
```

Through this template, we can know the elements of the sql to be generated, such as the name of the table, the name of the database, and various attributes of the table. It is easier to read this part of the code.

danny0405 replied:
Not a fan of antlr, this introduces unnecessary complexity and dependency of the antlr jar, how about we refactor the code using just a string builder, each sub-clause can be built separately.

slfan1989 replied:
I agree with you. I will refactor this part of code with string builder.

slfan1989 replied:
@danny0405 Thank you very much for helping to review the code and providing suggestions for improvement! I have refactored this part of the code using StringBuilder, as we discussed, to improve its readability. I also added some comments to provide additional context and clarification. Sorry for the delayed response.

When you have time, could you please review this PR again? Thank you very much!

slfan1989 commented:
@danny0405 Can you help review this PR again? Thank you very much!

slfan1989 commented:
@danny0405 Thank you very much for helping to review the code and providing suggestions for improvement! I have refactored this part of the code using `StringBuilder`, as we discussed, to improve its readability. I also added some comments to provide additional context and clarification. Sorry for the delayed response.

When you have time, could you please review this PR again? Thank you very much!

slfan1989 commented:
@hudi-bot run

danny0405 commented:
Thanks for the contribution, I have reviewed and created a patch:
[6086.patch.zip](https://github.com/apache/hudi/files/11370752/6086.patch.zip)

I saw some sub-claused locations are changed, like the `LOCATION` and `CLUSTETERED BY`, is that as expected?

slfan1989 commented:
@danny0405 

Thank you very much for your reminder! I carefully readed the code again and modify the code. 

The concatenation order is consistent between the original and modified versions, both of which consist of `7` steps.
1. Append Create Table 
2. Append Columns
3. Append Partitions
4. Append Bucket
5. Append Row Format
6. Append Location
7. Append TblProperties

> Original Version.

![image](https://user-images.githubusercontent.com/55643692/235890395-08099afc-bcbb-4c00-923c-960c871ff490.png)

> Modified Version.

<img width="794" alt="image" src="https://user-images.githubusercontent.com/55643692/235891453-0194117f-7ca3-44f7-b097-e0440f1dc76e.png">

slfan1989 commented on hudi-sync/hudi-hive-sync/src/main/java/org/apache/hudi/hive/ddl/QueryBasedDDLExecutor.java:
```diff
@@ -76,22 +77,41 @@ public QueryBasedDDLExecutor(HiveSyncConfig config) {
+   * @param serdeClass serdeClass.
+   * @param serdeProperties serdeProperties.
+   * @param tableProperties tableProperties.
+   */
```
Thank you for your suggestion! I will simplify the comment information.

slfan1989 commented:
> I saw some sub-claused locations are changed, like the `LOCATION` and `CLUSTETERED BY`, is that as expected?

@danny0405 Thank you for your thorough review! The modification to the `LOCATION` code should be as expected, but there were some issues with the modification to the `BUCKET` code. I have reverted the `BUCKET` code back to its original state.

danny0405 commented:
> > I saw some sub-claused locations are changed, like the `LOCATION` and `CLUSTETERED BY`, is that as expected?
> 
> @danny0405 Thank you for your thorough review! The modification to the `LOCATION` code should be as expected, but there were some issues with the modification to the `BUCKET` code. I have reverted the `BUCKET` code back to its original state.

We better have a straight-forward string paradigm of what the DDL looks like before and after the change, in case there are some discrepencies.

slfan1989 commented:
I will add a set of unit tests to compare the generated SQL before and after modification.

danny0405 commented on hudi-sync/hudi-hive-sync/src/test/java/org/apache/hudi/hive/util/TestHiveSchemaUtil.java:
```diff
@@ -145,4 +145,11 @@ public void testSchemaDiffForTimestampMicros() {
+    String expectedCreateDataBaseSQL = "CREATE DATABASE IF NOT EXISTS `test_database`";
+    String testDataBase = HiveSchemaUtil.generateCreateDataBaseDDL("test_database");
+    assertEquals(expectedCreateDataBaseSQL, testDataBase);
+  }
```
Can we also have a test case for create table?

slfan1989 replied:
Thanks for your suggestion, I will submit the test code as soon as possible.

danny0405 commented on hudi-sync/hudi-hive-sync/src/main/java/org/apache/hudi/hive/ddl/QueryBasedDDLExecutor.java:
```diff
@@ -76,25 +77,37 @@ public QueryBasedDDLExecutor(HiveSyncConfig config) {
 
+  /**
+   * Create a table with the given params.
+   */
```
Is the doc correct?

danny0405 commented on hudi-sync/hudi-hive-sync/src/main/java/org/apache/hudi/hive/ddl/QueryBasedDDLExecutor.java:
```diff
@@ -113,6 +126,9 @@ public void updateTableDefinition(String tableName, MessageType newSchema) {
 
+  /**
+   * Create a table with the given params.
+   */
```
ditto

danny0405 commented on hudi-sync/hudi-hive-sync/src/main/java/org/apache/hudi/hive/util/HiveSchemaUtil.java:
```diff
@@ -446,58 +448,76 @@ public static String generateSchemaString(MessageType storageSchema, List<String
 
-  public static String generateSchemaString(MessageType storageSchema, List<String> colsToSkip, boolean supportTimestamp) throws IOException {
+  /**
+   * Generates the Column DDL string for creating a Hive table from the given schema.
```
I love the original document that we add a SQL statement demo for the method, can we keep that?
